File: native_grayscale_sdk/ios/NativeGrayscaleSDK/Sources/Core/NativeGrayscaleSDKError.swift
//
//  NativeGrayscaleSDKError.swift
//  NativeGrayscaleSDK
//
//  Created by younggi.lee
//

import Flutter
import Foundation

/// SDK 에러 코드 정의
public enum NativeGrayscaleSDKErrorCode: Int {
  // Method Channel 관련 에러 (1000번대)
  case unknownMethod = 1001
  case methodCallError = 1002
  
  // Method Channel 초기화 관련 에러 (2000번대)
  case methodChannelNotInitialized = 2001
  case invalidResponse = 2003
  
  // Image 처리 관련 에러 (1100번대)
  case invalidArgument = 1101
  case imagePathRequired = 1102
  case conversionError = 1103
  case imageFileNotFound = 1104
  
  // Log 관련 에러 (1200번대)
  case setLogInterceptorError = 1201
  case setMinimumLogLevelError = 1202
  
  // 기타 에러 (9000번대)
  case unknownError = 9001
  
  /// 기본 에러 메시지
  var defaultMessage: String {
    switch self {
    case .unknownMethod:
      return "Unknown method"
    case .methodCallError:
      return "Error handling method call"
    case .methodChannelNotInitialized:
      return "Method channel is not initialized"
    case .invalidResponse:
      return "Invalid response from Dart"
    case .invalidArgument:
      return "Invalid argument"
    case .imagePathRequired:
      return "imagePath is required"
    case .conversionError:
      return "Failed to convert image to grayscale"
    case .imageFileNotFound:
      return "Image file does not exist at path"
    case .setLogInterceptorError:
      return "Failed to set log interceptor"
    case .setMinimumLogLevelError:
      return "Failed to set minimum log level"
    case .unknownError:
      return "Unknown error occurred"
    }
  }
}

/// NativeGrayscaleSDK SDK 에러 정의
///
/// Dart에서 전달된 에러를 파싱하고, 네이티브 앱으로 전달할 수 있는 형식으로 변환합니다.
public struct NativeGrayscaleSDKError: Error, LocalizedError {
  /// 에러 코드
  public let code: Int
  
  /// 에러 메시지
  public let message: String
  
  public init(code: Int, message: String) {
    self.code = code
    self.message = message
  }
  
  /// 에러 코드 enum을 사용하여 에러 생성
  public init(code: NativeGrayscaleSDKErrorCode, message: String? = nil) {
    self.code = code.rawValue
    self.message = message ?? code.defaultMessage
  }
  
  /// Dart에서 전달된 에러 딕셔너리를 파싱하여 NativeGrayscaleSDKError를 생성합니다.
  ///
  /// - Parameter errorDict: {'error': {'code': 1001, 'message': "error desc"}} 형식의 딕셔너리
  /// - Returns: 파싱된 NativeGrayscaleSDKError, 파싱 실패시 nil
  public static func fromDictionary(_ errorDict: [String: Any]?) -> NativeGrayscaleSDKError? {
    guard let errorDict = errorDict,
          let error = errorDict["error"] as? [String: Any],
          let code = error["code"] as? Int,
          let message = error["message"] as? String else {
      return nil
    }
    return NativeGrayscaleSDKError(code: code, message: message)
  }
  
  /// FlutterError의 details에서 NativeGrayscaleSDKError를 추출합니다.
  ///
  /// - Parameter flutterError: FlutterError 인스턴스
  /// - Returns: 파싱된 NativeGrayscaleSDKError, 파싱 실패시 nil
  public static func fromFlutterError(_ flutterError: FlutterError) -> NativeGrayscaleSDKError? {
    if let details = flutterError.details as? [String: Any] {
      return fromDictionary(details)
    }
    // FlutterError의 message와 code를 사용하여 기본 에러 생성
    return NativeGrayscaleSDKError(
      code: .unknownError,
      message: flutterError.message ?? "Unknown error"
    )
  }
  
  public var errorDescription: String? {
    return message
  }
  
  public var failureReason: String? {
    return "Error code: \(code)"
  }
}

File: native_grayscale_sdk/flutter/lib/errors/native_grayscale_sdk_error.dart
import 'package:flutter/services.dart';

/// NativeGrayscaleSDK SDK 에러 정의
///
/// 모든 에러 케이스를 enum으로 정의하고, 네이티브로 전달할 수 있는 형식으로 변환합니다.
enum NativeGrayscaleSDKError {
  // Method Channel 관련 에러 (1000번대)
  unknownMethod(1001, 'Unknown method'),
  methodCallError(1002, 'Error handling method call'),

  // Image 처리 관련 에러 (1100번대)
  invalidArgument(1101, 'Invalid argument'),
  imagePathRequired(1102, 'imagePath is required'),
  conversionError(1103, 'Failed to convert image to grayscale'),
  imageFileNotFound(1104, 'Image file not found'),

  // Log 관련 에러 (1200번대)
  setLogInterceptorError(1201, 'Failed to set log interceptor'),
  setMinimumLogLevelError(1202, 'Failed to set minimum log level'),

  // 기타 에러 (9000번대)
  unknownError(9001, 'Unknown error occurred');

  final int code;
  final String defaultMessage;

  const NativeGrayscaleSDKError(this.code, this.defaultMessage);

  /// 에러를 네이티브로 전달할 수 있는 Map 형식으로 변환합니다.
  ///
  /// Returns: {'error': {'code': 1001, 'message': "error desc"}}
  Map<String, dynamic> toMap([String? customMessage]) {
    return {
      'error': {'code': code, 'message': customMessage ?? defaultMessage},
    };
  }

  /// 에러를 PlatformException으로 변환합니다.
  PlatformException toPlatformException([String? customMessage]) {
    return PlatformException(code: name, message: customMessage ?? defaultMessage, details: {'errorCode': code});
  }

  /// Map에서 NativeGrayscaleSDKError를 생성합니다.
  ///
  /// 주로 네이티브에서 받은 에러를 파싱할 때 사용합니다.
  static NativeGrayscaleSDKError? fromMap(Map<dynamic, dynamic>? errorMap) {
    if (errorMap == null) return null;

    final errorCode = errorMap['code'] as int?;
    if (errorCode == null) return null;

    return NativeGrayscaleSDKError.values.firstWhere(
      (e) => e.code == errorCode,
      orElse: () => NativeGrayscaleSDKError.unknownError,
    );
  }
}
